### PR TITLE
batched evaluation of evaluate_notranspose for SpinorSet

### LINKED ISSUE #3712: Commit plan for batched spin-orbit code
**Is your feature request related to a problem? Please describe.**
Below is a checklist to help me keep track of what is done for the SO capability in the batched code. This helps close #3508 

**Describe the solution you'd like**

- [x] add spin moves in multiwalker APIs (#3710, #3728)
- [x] add spin gradients (e.g. evalGradWithSpin APIs) to TWF and TWFdispatcher for mw evaluation (#3714, #3887)
- [x] add spin gradients (e.g. evalGradWithSpin APIs) to  DiracDeterminant for mw evaluation (#3887, #4182, #4208, #4269 )
- [x] add mw evaluation to SpinorSet (#3726, #3730, #4208, #4226, #4231)
- [x] update VMC/DMC batched drivers to include spin moves (#3774, #3762, #3867)
- [x] update SOECPComponent and SOECPotential to include mw evaluations(#4375, #4504)

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

This is the first mw_ function in SpinorSet. I wanted to implement this first before following up with the rest of the functions to keep it small and easier to review. 

This implements mw_evaluate_notranspose. Since a SpinorSet has unique_ptr<SPOSet> for both up and down SPOSets, in order to utilize the batched evaluations for the individual spin components I had to form intermediate reference vectors. After these are formed, I simply can call mw_evaluate_notranspose for each of the spin components. Afterwards, I finish up by reconstructing the spinors from the spin components. 

Unit test is updated, I simply add a second walker by permuting the  electrons in the original particleset so that I can use the same reference data. 

This will be one of the needed PRs to close #3712 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
macOSX

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCWaveFunctions/tests/test_einset_spinor.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/tests/test_einset_spinor.cpp:463]: I will be better to initialize shared resource for particleset to avoid potential issues with GPUs
https://github.com/QMCPACK/qmcpack/blob/908bcb018cfab78bf15295a881fb4e8f1c83d383/src/QMCWaveFunctions/tests/test_TrialWaveFunction.cpp#L208

Comment (camelto2) [reply src/QMCWaveFunctions/tests/test_einset_spinor.cpp:470]: Done. also added as well as the ResourceCollectionTeamLock which I saw was also used in test_TrialWaveFunction

File: src/QMCWaveFunctions/SpinorSet.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/SpinorSet.cpp:219]: No need to clear. `mw_up_logdet` defined in the scope start with size 0.

Comment (camelto2) [reply src/QMCWaveFunctions/SpinorSet.cpp]: fixed

**Review by ye-luo — APPROVED** [review]
LGTM. Let us see what happens with CI.

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Wait for a review from @rcclay

**Pull request merged**
